File: python/1583_统计不开心的朋友_medium.py
class Solution:
    def unhappyFriends(self, n: int, preferences: List[List[int]], pairs: List[List[int]]) -> int:
        sequence, match = [[0] * n for _ in range(n)], [0]*n
        ans = 0
        for i in range(n):
            for j in range(n-1):
                sequence[i][preferences[i][j]] = j
        for pair in pairs:
            match[pair[0]] = pair[1]
            match[pair[1]] = pair[0]

        for x in range(n):
            y = match[x]
            index = sequence[x][y]
            for i in range(index):
                u = preferences[x][i]
                v = match[u]
                if sequence[u][x] < sequence[u][v]:
                    ans += 1
                    break
        return ans

File: cpp/1583_统计不开心的朋友_medium.cpp
class Solution {
public:
    int unhappyFriends(int n, vector<vector<int>>& preferences, vector<vector<int>>& pairs) {
        vector<vector<int>> sequence(n, vector<int>(n));
        for (int i=0; i<n; i++)
            for (int j=0; j<n-1; j++)
                sequence[i][preferences[i][j]] = j;
        vector<int> match(n);
        for (int i=0; i<pairs.size(); i++) {
            int person0 = pairs[i][0], person1 = pairs[i][1];
            match[person0] = person1;
            match[person1] = person0;
        }
        int ans = 0;
        for (int x=0; x<n; x++) {
            int y = match[x];
            int index = sequence[x][y];
            for (int i=0; i<index; i++) {
                int u = preferences[x][i];
                int v = match[u];
                if (sequence[u][x] < sequence[u][v]) {
                    ans++;
                    break;
                }
            }
        }
        return ans;
    }
};